### PR TITLE
Fix that SCTs were not being correctly signed

### PR DESCRIPTION
**Pull request by Martin2112**
Also builds MerkleTreeLeaf structure for them, which will shortly be used in the add chain handlers.

### CODE REVIEW

File: examples/ct/structures.go [hunk 1]
```diff
@@ -60,21 +60,60 @@ func signSCT(km crypto.KeyManager, t time.Time, sctData []byte) (ct.SignedCertif
...
+	// This is only possible using the CT specific modified version of X.509.
 	keyHash := sha256.Sum256(cert.RawSubjectPublicKeyInfo)
-	tbsBytes := make([]byte, 0, len(cert.RawTBSCertificate)+sha256.Size)
-	tbsBytes = append(tbsBytes, keyHash[:]...)
-	tbsBytes = append(tbsBytes, cert.RawTBSCertificate...)
+	precert := ct.PreCert{IssuerKeyHash: keyHash, TBSCertificate: cert.RawTBSCertificate}
+
+	timestampedEntry := ct.TimestampedEntry{Timestamp:sctInput.Timestamp, EntryType:ct.PrecertLogEntryType, PrecertEntry:precert}
+	leaf := ct.MerkleTreeLeaf{Version: ct.V1, LeafType:ct.TimestampedEntryLeafType, TimestampedEntry:timestampedEntry}
+	logEntry := ct.LogEntry{Leaf:leaf}
```
Comment (phad) [inline examples/ct/structures.go]: logEntry was inlined in the function above

File: examples/ct/structures.go [hunk 2]
```diff
@@ -60,21 +60,60 @@ func signSCT(km crypto.KeyManager, t time.Time, sctData []byte) (ct.SignedCertif
...
+	// Create a complete SCT including signature
+	sct, err := signSCT(km, t, res)
+
+	return leaf, sct, err
 }
 
 // CreateV1SCTForPrecertificate builds and signs a V1 CT SCT for a pre-certificate using the key
 // held by a key manager.
-func SignV1SCTForPrecertificate(km crypto.KeyManager, cert *x509.Certificate, t time.Time) (ct.SignedCertificateTimestamp, error) {
+func SignV1SCTForPrecertificate(km crypto.KeyManager, cert *x509.Certificate, t time.Time) (ct.MerkleTreeLeaf, ct.SignedCertificateTimestamp, error) {
```
Comment (phad) [inline examples/ct/structures.go]: there's a lot of duplicated code.

can you extract a helper function that does the bulk of it, and to which these exported functions pass a func object for the bits in the middle that differ?

File: examples/ct/structures.go (hunk 1 above)
Comment (Martin2112) [inline examples/ct/structures.go]: Fixed.

File: examples/ct/structures.go (hunk 2 above)
Comment (Martin2112) [inline examples/ct/structures.go]: Inspiration taken from cpp/log/log_signer.cc. I've stuck in an extra level to cut the duplication down.

**Pull request merged**